### PR TITLE
 Added MBTiles support for iOS and Android Vol. II

### PR DESCRIPTION
**Pull request by Stophface**
I submitted this `PR` already, but something went wrong with merging it into the master. Please refer to this https://github.com/react-community/react-native-maps/pull/2208 for the discussion. In this `PR`, I added the line `import android.location.Location;` in `AirMapView.java` which should prevent breaking the compiling.

### CODE REVIEW

**Comment by stobis** [comment]
There would be (probably) still a problem with `setKmlSrc` function, used in `AirMapManager.java` but unfortunately missing...

**Comment by Stophface** [comment]
@stobis Then someting went wrong when I merged the master into my branch? I did not edit with these lines in this file...

**Comment by stobis** [comment]
Honestly I don't know :/ I'm not familiar with this project, but that's possible.

**Comment by Stophface** [comment]
I will add the missing `setKmlSrc`.

**Comment by rborn** [comment]
@alvelig I think will need your help here, https://github.com/react-community/react-native-maps/pull/2388/files#diff-aa0115c2f9965ac7541b6f71de28b024 looks really weird to me now, `onPoiClick` is deleted, lot of KML related imports are removed  and `setKmlSrc` has to be fixed so I wonder if it's me or something went awfully wrong?

**Comment by Stophface** [comment]
@rborn Were the lines added before or after the release 0.20.1

**Comment by rborn** [comment]
@Stophface nope :(

**Comment by h3ll0w0rld123** [comment]
Looks like it's https://github.com/Stophface/react-native-maps/commit/566bf0f3f7ef5b9766e95253beb4a66bb67f249d#diff-aa0115c2f9965ac7541b6f71de28b024, I think you accidentally removed a bunch of things, @Stophface? 

Sorry to resurface this and sorry to be out of the loop in a while, but my understanding is we have some unknown removals. It looks like the commit above removed a bunch of things. Is it the lack of test coverage that we're not seeing it?

**Comment by h3ll0w0rld123** [comment]
@Stophface do you remember why you removed those lines? Can't you just put those lines back?

**Comment by Stophface** [comment]
@h3ll0w0rld123 I see that these lines got removed during a commit from me, but I am more than 100% sure that I did not do this manually! Maybe that happened during a merge or whatever, but I am 1000% postive that this was not my doing. Maybe it has something to do that I cloned the repo at version `0.20.1` and did my changes ontop of that. 

Weirdly, when you use this project https://github.com/Stophface/maps which links to this repository https://github.com/Stophface/react-native-maps-0.20.1 everything works just fine. 

I manually copy & pasted the changed files from this repo https://github.com/Stophface/react-native-maps-0.20.1 into this https://github.com/Stophface/react-native-maps which I am trying to pull into the original now. 

As I said, I am very sure that I did not change these files manually. Maybe that happend during some weird merge or whatever. Maybe, sometime, I will look into that. But now - to be honest - I do not feel like it.

File: lib/android/src/main/java/com/airbnb/android/react/maps/AirMapView.java [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import android.graphics.Point;
 import android.graphics.PorterDuff;
 import android.os.Build;
+import android.os.Handler;
```
Comment (h3ll0w0rld123) [inline lib/android/src/main/java/com/airbnb/android/react/maps/AirMapView.java]: @rborn It looks like this file has nothing to do with the feature. Would it be correct just to revert this?

**Comment by h3ll0w0rld123** [comment]
@Stophface Let me know if you want me to take this over. I'm excited that at least the maintainers are on board with the feature, to me it seems like only that file has to be reverted.

If you don't want to do it I can take it over, I just don't know how to build and test the changes. There are 2 things we can do. You can add me as a collaborator to your fork and I can make changes while keeping your work, or I can fork it from the main and apply your changes. The former is preferred since it will give you credit for the work. 

Let me know, I really want to push this along because I'd rather deal with one mbtile file then a bunch of image files. This is a lot more elegant of a solution, both in terms of maintenance and functionality. We're really close, unless I'm mistaken there's only one file that has changes that it shouldn't have.

If you don't know how to revert a file, I can show you:
1. `$ git checkout origin/master lib/android/src/main/java/com/airbnb/android/react/maps/AirMapView.java` (I believe your master should be the same as the original project's). 
2. Literally just open the original project's corresponding AirMapView file and copy and paste it into your file.

**Comment by Stophface** [comment]
@h3ll0w0rld123 If that is the only file that needs changing, I will revert it tonight.

**Comment by h3ll0w0rld123** [comment]
@Stophface When I looked at it without much knowledge it seems to be the only thing that had weird changes. Checked with the other comments from the last PR about weird things, and it seems like it was isolated to that file.

Maintainers might have more context/knowledge?

**Comment by h3ll0w0rld123** [comment]
@Stophface ping?

**Comment by Stophface** [comment]
@h3ll0w0rld123 I copy & pasted the code from the master, but included

```
else if (child instanceof AirMapMbTile) {
      AirMapMbTile mbTileView = (AirMapMbTile) child;
      mbTileView.addToMap(map);
      features.add(index, mbTileView);
    }
```

Otherwise the component would not be accessable.

**Comment by h3ll0w0rld123** [comment]
@rborn  @alvelig Can we merge now?

**Comment by h3ll0w0rld123** [comment]
Ping @rborn @alvelig?

**Comment by rborn** [comment]
@h3ll0w0rld123 , sorry I was away, can you confirm that the PR works correctly? Can you run a test?

Also is the https://github.com/react-community/react-native-maps/pull/2388#issuecomment-410442430 present in the PR ?


🤗

**Comment by h3ll0w0rld123** [comment]
@rborn, my apologies, it looks like nothing changed. I'll take a look tomorrow and if I can set everything up easily, I'll do my own pull request separate from this.

**Comment by goliath45** [comment]
Hi

**Comment by Stophface** [comment]
@h3ll0w0rld123 @rborn The changes are in there?

https://github.com/Stophface/react-native-maps/blob/feature/mbtiles/lib/android/src/main/java/com/airbnb/android/react/maps/AirMapView.java

I reverted so `import android.location.Location;` as well as ` public void setKmlSrc(String kmlSrc) {...` is in there?!

**Comment by h3ll0w0rld123** [comment]
@Stophface can you update the README.md, and @rborn can you check to see if it's mergeable?

**Comment by h3ll0w0rld123** [comment]
@Stophface https://github.com/react-community/react-native-maps/pull/2425/files is the change that caused a conflict. If you give me write permissions on your repo I can reconcile it.

**Comment by Stophface** [comment]
@h3ll0w0rld123 You should have an invitation.

**Comment by h3ll0w0rld123** [comment]
@rborn @alvelig the problematic file seems to have been fixed and it's mergeable. Can you guys check and merge if all is okay?

I tried to test it myself, but since I develop using Expo I'm kinda out of my element here.

I tried building this into a tarball
```
npm pack
```
then including it in a new app
```
$ react-native init testApp
$ cd testApp
$ npm install /path/to/tarball/react-native-maps-0.21.0.tgz
$ react-native run-ios
```
But I got some file errors when running.

**Comment by h3ll0w0rld123** [comment]
@rborn @alvelig so sorry for the bother, but just pinging again :)

**Comment by simon4321** [comment]
Any updates on this?

**Comment by storerjeremy** [comment]
@h3ll0w0rld123 @Stophface It looks like your [test app](https://github.com/Stophface/maps) is using rastor mbtiles, does this pull request provide support for vector mbtiles as well?

**Comment by Stophface** [comment]
@storerjeremy Only raster.

**Comment by jgtoriginal** [comment]
is this PR likely to get merged anytime soon?

**Comment by vnlebaoduy** [comment]
Please check PR :( . I need this PR @Stophface

**Comment by Stophface** [comment]
Sorry, I do not have time to look into this...

**Comment by secretgh** [comment]
This is an awesome functionality to have is there any way to move this along?

**Comment by madhums** [comment]
@Stophface any chance you would be able to rebase this? I'd love to use what you have developed!

Also, what needs to be done in order to merge this PR? @rborn @Stophface

**Comment by madhums** [comment]
I managed to rebase, you can find the branch here https://github.com/madhums/react-native-maps/tree/feature/mbtiles. 

When using this, I run into this error unfortunately https://github.com/react-native-community/react-native-maps/issues/2369 

Screenshot 2019-12-30 at 21 12 48

It's not explained properly in that issue the reason for this error (everyone's busy posting what solution worked for them). To be clear, I am using it with expo@36.0.0 without ejecting. Does anyone know the reason for this error and what it means?

**Comment by secretgh** [comment]
@madhums 
I ran into this issue as well, when you rebase there are changes that were made that removed all mention of the all mbtile components, you need to update AIRMapManager and the MapManager to have mentioned the correct components (AirMapMBTile and MBTile) I don't remember the exact names, as I'm not close to my computer right now but just follow a similar code as the UrlMapTile

**Comment by madhums** [comment]
@secretgh yea I've double checked it many times already now. Those files do reflect the changes as made in the original PR. Here you [can compare](https://github.com/react-native-community/react-native-maps/compare/master...madhums:feature/mbtiles)

**Comment by madhums** [comment]
anybody?

**Comment by madhums** [comment]
I managed to run the example but having some inconsistencies in the map data. Not sure what is the reason for this. 

Here's [sample mbtiles](https://gitlab.com/IvanSanchez/Leaflet.TileLayer.MBTiles/blob/master/demo/countries-raster.mbtiles) I am using.

Online:
Screenshot 2020-01-21 at 13 16 24

Offline: (loaded using mbtiles)
Screenshot 2020-01-21 at 13 16 20

Is this something to do with EPSG? Should I be doing any conversions?

**Comment by madhums** [comment]
Umm wondering if this might be a problem with the delta value I am using.

**Side note:**
It looks like there's numerous assignments of delta values without any explanation of how it came to be. For example search in this repo for [0.0922](https://github.com/react-native-community/react-native-maps/search?q=0.0922&unscoped_q=0.0922). There are issues like #505 #3200 which is probably caused by this poor developer experience which probably could be fixed by giving a proper explanation or formula for deriving these values.

**Comment by madhums** [comment]
Ok so it turns out to be this issue that is blocking. Even after ejecting from Expo it still throws [this error](https://github.com/react-native-community/react-native-maps/pull/2388#issuecomment-569784551)

```
Invariant violation: requireNativeComponent: "AIRMapMbTile" was not found in the UIManager"
```

Anybody have any clue how to get past this?

**Comment by SaeedZhiany** [comment]
a

**Comment by christopherdro** [comment]
@Stophface @h3ll0w0rld123 @rborn Thoughts on reviving this PR?

**Pull request closed without merging**